feat(corebase): Add core-base/datastore module

HekmatullahAmin opened the pull request:
Fixes - [Jira-#KMPPT-12](https://mifosforge.jira.com/browse/KMPPT-12)


## Demo

https://github.com/user-attachments/assets/bb024afb-ceb6-456c-9391-c2bc7b23dac5

https://github.com/user-attachments/assets/06907b1b-e447-4bbf-973e-aee4dbe2be3d

https://github.com/user-attachments/assets/f0749ea7-a975-4a39-ae33-a6b2e92b8732

HekmatullahAmin committed:
feat(corebase): Add core-base/datastore module

HekmatullahAmin commented:
<img src="https://github.com/user-attachments/assets/65ae701b-8bf4-47f7-890d-4ee29b355ea1" width="50" alt="red-cross"> This initial documentation reflects the use of SuspendSettings and FlowSettings. However, I refactored it because these only work for Android, iOS, and JVM—not Web. I’ve switched to Settings for full cross-platform support. Scroll down for the latest implementation documentation.


## What is Achieved
I’ve moved all the saving and retrieving logic into a centralized UserPreferencesDataStore class, making it much easier to manage preferences across the app. You no longer need to worry about the underlying details of storing or fetching data—whether it’s primitive types or serialized objects. Everything is now abstracted into three simple functions: `putValue`, `getValue`, and `observeKeyFlow`.

What You Need to Do
You can now use these three functions for all your data storage needs. Here’s how they work, along with examples:

# `putValue` - Save Any Value
Purpose: Saves a value under a given key, handling both primitive types and serialized data for you.
Structure:
```
suspend fun <T> putValue(
    key: String,
    value: T,
    serializer: KSerializer<T>? = null
)
```

How to Use:
- For primitive types (e.g., Int, Double), just pass the key and value.
- For serialized data (e.g., custom data classes), also pass a serializer.
Example:
```
// Saving a primitive Double
val dataStore = UserPreferencesDataStore(suspendSettings, flowSettings)
runBlocking {
    dataStore.putValue("user_balance", 99.99)
}

// Saving a serialized data class
@Serializable
data class User(val name: String, val age: Int)

val user = User("Alice", 25)
runBlocking {
    dataStore.putValue("user_profile", user, User.serializer())
}
```

# getValue - Retrieve Any Value
Purpose: Retrieves a value for a given key, returning a default value if the key isn’t found. Supports primitive types natively and custom types via serialization.
Structure:
```
suspend fun <T> getValue(
    key: String,
    default: T,
    serializer: KSerializer<T>? = null
)
```

How to Use:
- For primitive types, pass the key and a default value.
- For serialized data, also pass a serializer.
Example:
```
// Retrieving a primitive Double
val dataStore = UserPreferencesDataStore(suspendSettings, flowSettings)
runBlocking {
    val balance = dataStore.getValue("user_balance", 0.0)
    println("Balance: $balance") // Prints stored value or 0.0 if not found
}

// Retrieving a serialized data class
@Serializable
data class User(val name: String, val age: Int)

runBlocking {
    val user = dataStore.getValue("user_profile", User("Default", 0), User.serializer())
    println("User: $user") // Prints stored User or default if not found
}
```

# observeKeyFlow - Observe Value Changes
Purpose: Observes changes to a value for a given key as a Flow, emitting the current value and any updates. Works for both primitive and serialized types.
Structure:
```
fun <T> observeKeyFlow(
    key: String,
    defaultValue: T,
    serializer: KSerializer<T>? = null
): Flow<T>
```

How to Use:
- For primitive types, pass the key and a defaultValue.
- For serialized data, also pass a serializer.
Example:
```
// Observing a primitive Double
val intNumber: Flow<Int> =
        dataStore.observeKeyFlow(key = INT_KEY, defaultValue = 0)

// Observing a serialized data class
val currentUser: Flow<SampleUser> =
        dataStore.observeKeyFlow(USER_KEY, SampleUser.DEFAULT, SampleUser.serializer())
```

HekmatullahAmin committed:
remove unnecessary dependencies

niyajali commented on core-base/datastore/src/commonMain/kotlin/org/mifos/corebase/datastore/di/CoreDatastoreModule.kt:
```diff
@@ -0,0 +1,47 @@
+val CoreDatastoreModule = module {
+    single { Settings() }
+    single {
+        SettingsFactory.createSuspendSettings(
```
What does **createSuspendSettings** mean? I don’t understand what it does or why we need to create something like this.

HekmatullahAmin replied:
@niyajali In my previous implementation, you might notice `createSuspendSettings` or `createFlowSettings`. Mr. @therajanmaurya  had advised using them, but after implementation, I found they only work for Android, iOS, and JVM—not Web. To ensure cross-platform support, I refactored it to use Settings instead, as it works across all platforms, including Web.

<img width="855" alt="Screenshot 2025-03-08 at 11 09 06" src="https://github.com/user-attachments/assets/13cc247d-859f-48a5-b88a-6d825c58f183" />

niyajali commented on core-base/datastore/src/commonMain/kotlin/org/mifos/corebase/datastore/di/CoreDatastoreModule.kt:
```diff
@@ -0,0 +1,47 @@
+        SettingsFactory.createFlowSettings(
+            settings = get(),
+            dispatcher = get(named(AppDispatchers.IO.name)),
+        )
```
Same here, what we trying to achieve

niyajali commented on core-base/datastore/src/commonMain/kotlin/org/mifos/corebase/datastore/UserPreferencesDataStore.kt:
```diff
@@ -0,0 +1,215 @@
+@OptIn(ExperimentalSettingsApi::class)
+class UserPreferencesDataStore(
+    private val suspendSettings: SuspendSettings,
+    private val flowSettings: FlowSettings,
```
Why do we Injected different settings instances.

niyajali commented on core-base/datastore/src/commonMain/kotlin/org/mifos/corebase/datastore/SettingsFactory.kt:
```diff
@@ -20,6 +20,7 @@ import kotlinx.coroutines.CoroutineDispatcher
 @OptIn(ExperimentalSettingsApi::class)
 object SettingsFactory {
+
     fun createSuspendSettings(
```
Where do we have these extension fun and why need this using the settings instances and ioDispatcher will sufficient no need to create such

niyajali commented on core-base/datastore/src/commonMain/kotlin/org/mifos/corebase/datastore/UserPreferencesDataStore.kt:
```diff
@@ -0,0 +1,215 @@
+     * @throws IllegalArgumentException If [default] is not a supported primitive type and [serializer] is null.
+     */
+    suspend fun <T> getValue(
+        key: String,
```
And i think these generic functions are already provided by settings library

HekmatullahAmin committed:
Refactor UserPreferencesDataStore to use Settings instead of SuspendSettings for cross-platform compatibility

HekmatullahAmin commented:
# What is Achieved
I’ve moved all the saving and retrieving logic into a centralized UserPreferencesDataStore class, making it much easier to manage preferences across the app. You no longer need to worry about the underlying details of storing or fetching data—whether it’s primitive types or serialized objects. Everything is now abstracted into two simple functions: putValue, getValue.

What You Need to Do
Use putValue and getValue from UserPreferencesDataStore to store and fetch data. Here’s how they work with examples:

## putValue - Save Any Value
Purpose: Saves a value under a key, handling both primitive types and custom objects with a serializer.

Structure:
```
suspend fun <T> putValue(key: String, value: T, serializer: KSerializer<T>? = null)
```
How to Use:

- For primitives (e.g., Double), pass the key and value.
- For custom objects (e.g., SampleUser), add a serializer.
Examples:
```
// Saving a Double
suspend fun saveDouble() {
    dataStore.putValue("sample_double", 99.99)
}

// Saving a SampleUser
@Serializable
data class SampleUser(val name: String, val age: Int)
suspend fun saveUser() {
    val user = SampleUser("Alice", 25)
    dataStore.putValue("sample_user", user, SampleUser.serializer())
}
```

## getValue - Retrieve Any Value
Purpose: Fetches a value by key, returning a default if not found. Supports primitives and custom objects with a serializer.

Structure:
```
suspend fun <T> getValue(key: String, default: T, serializer: KSerializer<T>? = null): T
```
How to Use:

- For primitives, pass the key and default value.
- For custom objects, include a serializer.
Examples:
```
// Retrieving a Double
suspend fun getDouble() {
    val balance = dataStore.getValue("sample_double", 0.0)
    println("Balance: $balance") // Prints stored value or 0.0
}

// Retrieving a SampleUser
@Serializable
data class SampleUser(val name: String, val age: Int)
suspend fun getUser() {
    val user = dataStore.getValue("sample_user", SampleUser("Default", 0), SampleUser.serializer())
    println("User: $user") // Prints stored user or default
}
```

HekmatullahAmin commented:
@niyajali Could you please review. I revert it back from suspendSettings and FlowSettings to Settings cause it is working for all platforms. If it is okay merge it.

niyajali commented:
Checks are failing fix that, and let us know we'll proceed with merge

HekmatullahAmin committed:
Removed SettingsFactory and updated ci-prepush.sh to use :cmp-android:build and :cmp-android:updateProdReleaseBadging

HekmatullahAmin commented:
Hi @niyajali ,
The PR is ready.

And also one of the reasons that checks were failing was that in ci-prepush.sh, the tasks `:mifos-android:build` and `:mifos-android:updateProdReleaseBadging` were being used, but those tasks don't exist in the project. So, I replaced them with `:cmp-android:build` and `:cmp-android:updateProdReleaseBadging`, which are the correct ones.

However, while running the build, I encountered the following error:

"Task ':cmp-android:detekt' uses this output of task ':cmp-android:dependencyGuard' without declaring an explicit or implicit dependency."

This was causing incorrect execution order, so I had to explicitly define:
```
tasks.named("detekt") {
    mustRunAfter(":cmp-android:dependencyGuard")
}
```
to ensure the tasks run in the correct sequence.

For reference, I’ve added the photos/screenshots.
<img width="1000" alt="Screenshot 2025-03-10 at 10 13 51" src="https://github.com/user-attachments/assets/369c0479-d238-4774-9401-6a52a120bc9f" />

<img width="500" alt="Screenshot 2025-03-10 at 10 17 30" src="https://github.com/user-attachments/assets/508a38e5-a5cf-4103-8fa5-22ca0bf0141c" />
